File: web/src/scripts/types/api/album.ts
import type { TimelineItem } from '@/scripts/types/generated/timeline'
import type {
  CameraSettings,
  TimeDetails,
  VisualAnalysis,
  Weather,
} from '@/scripts/types/api/fullPhoto.ts'
import type { PannellumConfig } from '@/scripts/types/api/pannellumConfig.ts'

// Enums
export type AlbumRole = 'Owner' | 'Contributor' | 'Viewer'
export type SortDirection = 'asc' | 'desc'
export type AlbumSortField = 'updatedAt' | 'latestPhoto' | 'name'
export type AlbumSort = 'DateAsc' | 'DateDesc' | 'AddedAsc' | 'AddedDesc' | 'None'

// Entities
export interface Album {
  id: string
  ownerId: number
  name: string
  description: string | null
  thumbnailId: string | null
  isPublic: boolean
  createdAt: string // ISO Date string
  updatedAt: string // ISO Date string
  mediaCount: number
  latestMediaItemTimestamp: string | null // ISO Date string
  earliestMediaItemTimestamp: string | null // ISO Date string
  sortMode: AlbumSort
}

export interface AlbumCollaborator {
  id: number
  albumId: string
  userId: number
  role: AlbumRole
  addedAt: string // ISO Date string
}

export interface AlbumMediaItemSummary {
  mediaItem: TimelineItem
  addedAt: string // ISO Date string
}

export interface AlbumSummary {
  name: string
  description: string | null
  relativePaths: string[]
}

// Request Payloads
export interface CreateAlbumRequest {
  name: string
  description?: string
  isPublic: boolean
  mediaItemIds: string[]
}

export interface UpdateAlbumRequest {
  name?: string
  isPublic?: boolean
  description?: string | null
  thumbnailId?: string | null
}

export interface AddMediaToAlbumRequest {
  mediaItemIds: string[]
}

export interface AddCollaboratorRequest {
  userId: number
  role: AlbumRole
}

export interface CheckInviteRequest {
  token: string
}

export interface AcceptInviteRequest {
  token: string
  name: string
  description?: string
}

export interface SharedMediaFeatures {
  mime_type: string
  size_bytes: number
  is_motion_photo: boolean
  motion_photo_presentation_timestamp?: number
  is_hdr: boolean
  is_burst: boolean
  burst_id?: string
  capture_fps?: number
  video_fps?: number
  is_nightsight: boolean
  is_timelapse: boolean
  audio_format?: string
  audio_channels?: number
  audio_sample_rate?: number
  compressor_id?: string
}

export interface SharedMediaItem {
  id: string
  filename: string
  width: number
  height: number
  is_video: boolean
  duration_ms: number | null
  taken_at_local: string // NaiveDateTime
  taken_at_utc: string | null // DateTime<Utc>
  timezone_name: string | null
  timezone_offset_seconds: number | null
  panorama_config: PannellumConfig
  use_panorama_viewer: boolean
  has_thumbnails: boolean
  visual_analyses: VisualAnalysis[]
  time: TimeDetails
  weather: Weather | null
  media_features: SharedMediaFeatures
  camera_settings: CameraSettings
  user_caption: string | null
  gps: null
}

export interface BackupInfo {
  filename: string
  sizeBytes: number
  createdAt: string // ISO Date string
}

File: server/crates/libs/common_services/src/database/tables/album/album.rs
use crate::api::album::interfaces::AlbumSort;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::fmt;
use std::fmt::Display;

#[derive(Debug, Serialize, Deserialize, sqlx::Type, PartialEq, Eq)]
#[sqlx(type_name = "album_role", rename_all = "lowercase")]
pub enum AlbumRole {
    Owner,
    Contributor,
    Viewer,
}

impl Display for AlbumRole {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = match self {
            Self::Owner => "owner",
            Self::Contributor => "contributor",
            Self::Viewer => "viewer",
        };
        f.write_str(s)
    }
}

/// Represents a single album in the database.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Album {
    pub id: String,
    pub owner_id: i32,
    pub name: String,
    pub thumbnail_id: Option<String>,
    pub description: Option<String>,
    pub is_public: bool,
    pub sort_mode: AlbumSort,
    pub media_count: i32,
    pub latest_media_item_timestamp: Option<DateTime<Utc>>,
    pub earliest_media_item_timestamp: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct AlbumSummary {
    pub name: String,
    pub description: Option<String>,
    pub relative_paths: Vec<String>,
}
